language/move-prover/bytecode/src/dataflow_domains.rs

```diff
@@ -0,0 +1,285 @@
+// Map Type
+
+#[derive(Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub struct MapDomain<K: Ord, V: AbstractDomain>(OrdMap<K, V>);
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> Default for MapDomain<K, V> {
+    fn default() -> Self {
+        Self(OrdMap::default())
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> From<OrdMap<K, V>> for MapDomain<K, V> {
+    fn from(ord_map: OrdMap<K, V>) -> Self {
+        Self(ord_map)
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> AsRef<OrdMap<K, V>> for MapDomain<K, V> {
+    fn as_ref(&self) -> &OrdMap<K, V> {
+        &self.0
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> Borrow<OrdMap<K, V>> for MapDomain<K, V> {
+    fn borrow(&self) -> &OrdMap<K, V> {
+        self.as_ref()
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> Deref for MapDomain<K, V> {
+    type Target = OrdMap<K, V>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> DerefMut for MapDomain<K, V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<K: Ord + Clone + Debug, V: AbstractDomain + Clone + Debug> Debug for MapDomain<K, V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.0)
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> std::iter::FromIterator<(K, V)>
+    for MapDomain<K, V>
+{
+    fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
+        let mut s = MapDomain::default();
+        for (k, v) in iter {
+            s.insert(k, v);
+        }
+        s
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> std::iter::IntoIterator for MapDomain<K, V> {
+    type Item = (K, V);
+    type IntoIter = im::ordmap::ConsumingIter<(K, V)>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> AbstractDomain for MapDomain<K, V> {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        let mut change = JoinResult::Unchanged;
+        for (k, v) in other.iter() {
+            change = change.combine(self.insert_join(k.clone(), v.clone()));
+        }
+        change
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> From<BTreeMap<K, V>> for MapDomain<K, V> {
+    fn from(m: BTreeMap<K, V>) -> Self {
+        m.into_iter().collect()
+    }
+}
+
+impl<K: Ord + Clone, V: AbstractDomain + Clone> MapDomain<K, V> {
+    /// Construct a singleton map.
+    pub fn singleton(k: K, v: V) -> MapDomain<K, V> {
+        (ordmap! {k => v}).into()
+    }
+
+    /// Join `v` with self[k] if `k` is bound, insert `v` otherwise
+    pub fn insert_join(&mut self, k: K, v: V) -> JoinResult {
+        let mut change = JoinResult::Unchanged;
+        self.0
+            .entry(k)
+            .and_modify(|old_v| {
+                change = change.combine(old_v.join(&v));
```

Comment:
I think

```suggestion
                change = old_v.join(&v);
```

is equivalent/slightly simpler.

Reply:
Done